### PR TITLE
feat(data/fintype): add choose_unique and construct inverses to bijections

### PR DESCRIPTION
**Pull request by sebzim4500**
TO CONTRIBUTORS:

Make sure you have:

 * [ ] reviewed and applied the coding style: [coding](./docs/style.md), [naming](./docs/naming.md)
 * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](./docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](./tests/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [ ] make sure definitions and lemmas are not redundant

For reviewers: [code review check list](./docs/code-review.md)

### CODE REVIEW

**Comment by cipher1024** [comment]
I think you PR-ed to the wrong repository. Try leanprover/mathlib

File: data/fintype.lean [hunk 1]
```diff
@@ -586,4 +586,59 @@ lemma fintype.card_equiv [fintype α] [fintype β] (e : α ≃ β) :
...
+end equiv
+
+section choose_unique
+open fintype
+open equiv
+
+variables [fintype α] [decidable_eq α] (p : α → Prop) [decidable_pred p]
+
+def choose_unique_x (p_exists : ∃ a : α, p a) (p_unique : ∀ a b : α, p a → p b → a = b) : {a // p a} := begin
+apply @trunc.lift (α ≃ fin (card α)) _ _ _ (equiv_fin α),
```
Comment (cipher1024) [inline data/fintype.lean:598]: indent this begin / end block

File: data/fintype.lean [hunk 2]
```diff
@@ -586,4 +586,59 @@ lemma fintype.card_equiv [fintype α] [fintype β] (e : α ≃ β) :
...
+intro equiv_fin,
+let G : fin (card α) → Prop := λ f, p (equiv_fin.inv_fun f),
+have G_exists : {f : fin (card α) // G f},
+apply @encodable.choose_x (fin (card α)) G,
```
Comment (cipher1024) [inline data/fintype.lean:602]: Enclose the proof of this `have` in curly brackets

**Comment by digama0** [comment]
You can push this branch to leanprover-community if you like (you have push access), although it should be in its own branch - the leanprover-community/master branch just tracks mathlib/master, so we would want this PR'd to mathlib instead, like Simon said.

File: data/fintype.lean [hunk 3]
```diff
@@ -5,7 +5,7 @@ Author: Mario Carneiro
 
 Finite types.
 -/
-import data.finset algebra.big_operators data.array.lemmas data.vector2
+import data.finset algebra.big_operators data.array.lemmas data.vector2 data.equiv.basic data.equiv.encodable
```
Comment (bryangingechen) [inline data/fintype.lean:8]: The `data.equiv.basic` import is unnecessary; it's imported via a dependency of `data.equiv.encodable`.

File: data/fintype.lean [hunk 4]
```diff
@@ -586,4 +586,59 @@ lemma fintype.card_equiv [fintype α] [fintype β] (e : α ≃ β) :
...
+have G_exists : {f : fin (card α) // G f},
+apply @encodable.choose_x (fin (card α)) G,
+cases p_exists with p_w p_h,
+existsi equiv_fin.to_fun p_w,
+change p (equiv_fin.inv_fun (equiv_fin.to_fun p_w)),
+rw equiv_fin.left_inv, assumption,
+cases G_exists, existsi (equiv_fin.inv_fun G_exists_val),
+exact G_exists_property,
+intros,
+have all_equal : (Π a b : {a // p a}, a = b),
```
Comment (cipher1024) [inline data/fintype.lean:610]: same for this `have`

File: data/fintype.lean [hunk 5]
```diff
@@ -586,4 +586,59 @@ lemma fintype.card_equiv [fintype α] [fintype β] (e : α ≃ β) :
...
+:= choose_unique (λ a, f a = b) (f_bij.right b) (λ a1 a2 b1 b2, by apply f_bij.left; rw [b1, b2])
+
+def bij_inv_left_inverse (f_bij : bijective f) : left_inverse (bij_inv f_bij) f
+:= λ a, f_bij.left (choose_unique_spec (λ a', f a' = f a))
+
+def bij_inv_right_inverse (f_bij : bijective f) : right_inverse (bij_inv f_bij) f
+:= λ b, choose_unique_spec (λ a', f a' = b)
+
+def bij_inv_bijective (f_bij : bijective f) : bijective (bij_inv f_bij)
+:= ⟨injective_of_left_inverse (bij_inv_right_inverse _)
```
Comment (cipher1024) [inline data/fintype.lean:641]: `:=` should be on the previous line.

File: data/fintype.lean (hunk 1 above)
Comment (digama0) [reply data/fintype.lean:598]: Also the `∃!` exists unique notation is available for this hypothesis

File: data/fintype.lean [hunk 6]
```diff
@@ -586,4 +586,59 @@ lemma fintype.card_equiv [fintype α] [fintype β] (e : α ≃ β) :
...
\ No newline at end of file
+end equiv
+
+section choose_unique
+open fintype
+open equiv
+
+variables [fintype α] [decidable_eq α] (p : α → Prop) [decidable_pred p]
+
+def choose_unique_x (p_exists : ∃ a : α, p a) (p_unique : ∀ a b : α, p a → p b → a = b) : {a // p a} := begin
```
Comment (cipher1024) [inline data/fintype.lean:597]: put `begin` on next line

Comment (bryangingechen) [reply data/fintype.lean:597]: The formatting recommendations for mathlib are in [the style guide](https://github.com/leanprover/mathlib/blob/master/docs/style.md).

File: data/fintype.lean (hunk 3 above)
Comment (digama0) [reply data/fintype.lean:8]: `data.fintype` should not depend on `data.equiv.encodable`. You should put any theorems requiring this in `data.equiv.list`.

**Pull request closed without merging**